File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/impl/utils/PolarFirmwareUpdateUtils.swift
//  Copyright © 2024 Polar. All rights reserved.

import Foundation
import Zip

class PolarFirmwareUpdateUtils {
    static let FIRMWARE_UPDATE_FILE_PATH = "/SYSUPDAT.IMG"
    static let DEVICE_FIRMWARE_INFO_PATH = "/DEVICE.BPB"

    public class FwFileComparator {
        private static let SYSUPDAT_IMG = "SYSUPDAT.IMG"

        static func compare(_ file1: String, _ file2: String) -> ComparisonResult {
            if file1.contains(SYSUPDAT_IMG) {
                return .orderedDescending
            } else if file2.contains(SYSUPDAT_IMG) {
                return .orderedAscending
            } else {
                return .orderedSame
            }
        }
    }
    
    static func readDeviceFirmwareInfo(client: BlePsFtpClient, deviceId: String) async -> PolarFirmwareVersionInfo? {
        let request = Protocol_PbPFtpOperation.with {
            $0.command = .get
            $0.path = DEVICE_FIRMWARE_INFO_PATH
        }
        do {
            let serializedBytes = try request.serializedData()
            let response = try await client.request(serializedBytes)
            let proto = try Data_PbDeviceInfo(serializedBytes: response as Data)
            return PolarFirmwareVersionInfo(
                deviceFwVersion: devicePbVersionToString(pbVersion: proto.deviceVersion),
                deviceModelName: proto.modelName,
                deviceHardwareCode: proto.hardwareCode
            )
        } catch {
            BleLogger.error("Failed to request device info: \(deviceId), error: \(error)")
            return nil
        }
    }

    static func isAvailableFirmwareVersionHigher(currentVersion: String, availableVersion: String) -> Bool {
        let current = currentVersion.split(separator: ".").map { Int($0)! }
        let available = availableVersion.split(separator: ".").map { Int($0)! }

        for i in 0..<current.count {
            if available.count > i {
                if current[i] < available[i] {
                    return true
                } else if current[i] > available[i] {
                    return false
                }
            }
        }
        return available.count > current.count
    }

    static func unzipFirmwarePackage(zippedData: Data) -> [String: Data]? {
        let temporaryDirectory = FileManager.default.temporaryDirectory
        
        let zipFilePath = temporaryDirectory.appendingPathComponent(UUID().uuidString + ".zip")
        do {
            try zippedData.write(to: zipFilePath)

            let destinationURL = temporaryDirectory.appendingPathComponent(UUID().uuidString)

            try Zip.unzipFile(zipFilePath, destination: destinationURL, overwrite: true, password: nil)

            let contents = try FileManager.default.contentsOfDirectory(at: destinationURL, includingPropertiesForKeys: nil)
            guard !contents.isEmpty else {
                BleLogger.error("unzipFirmwarePackage() error: No files found in the extracted directory")
                return nil
            }
            var fileDataDictionary: [String: Data] = [:]
            for fileURL in contents {
                let fileName = fileURL.lastPathComponent
                let decompressedData = try Data(contentsOf: fileURL)
                fileDataDictionary[fileName] = decompressedData
                BleLogger.trace("Extracted file: \(fileName) - Size: \(decompressedData.count) bytes")
            }

            try FileManager.default.removeItem(at: zipFilePath)
            try FileManager.default.removeItem(at: destinationURL)

            return fileDataDictionary
        } catch {
            BleLogger.error("Error during unzipFirmwarePackage(): \(error)")
            return nil
        }
    }
    
    private static func devicePbVersionToString(pbVersion: PbVersion) -> String {
        return "\(pbVersion.major).\(pbVersion.minor).\(pbVersion.patch)"
    }
}

File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/impl/utils/PolarFirmwareUpdateUtils.kt
package com.polar.sdk.impl.utils

import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import com.polar.sdk.api.model.PolarFirmwareVersionInfo
import fi.polar.remote.representation.protobuf.Device
import fi.polar.remote.representation.protobuf.Structures
import protocol.PftpRequest
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.util.zip.ZipInputStream

internal object PolarFirmwareUpdateUtils {

    /**
     * Comparator for sorting FW files so that the order doesn't matter as long as
     * the SYSUPDAT.IMG file is the last one (since it makes the device boot itself).
     */
    class FwFileComparator : Comparator<File> {
        companion object {
            private const val SYSUPDAT_IMG = "SYSUPDAT.IMG"
        }

        override fun compare(f1: File, f2: File): Int {
            return when {
                f1.name.contains(SYSUPDAT_IMG) -> 1
                f2.name.contains(SYSUPDAT_IMG) -> -1
                else -> 0
            }
        }
    }

    const val FIRMWARE_UPDATE_FILE_PATH = "/SYSUPDAT.IMG"
    const val BUFFER_SIZE = 8192

    private const val DEVICE_FIRMWARE_INFO_PATH = "/DEVICE.BPB"
    private const val TAG = "PolarFirmwareUpdateUtils"

    suspend fun readDeviceFirmwareInfo(client: BlePsFtpClient, deviceId: String): PolarFirmwareVersionInfo {
        BleLogger.d(TAG, "readDeviceFirmwareInfo: $deviceId")
        val response = client.request(
            PftpRequest.PbPFtpOperation.newBuilder()
                .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                .setPath(DEVICE_FIRMWARE_INFO_PATH)
                .build()
                .toByteArray()
        )
        val proto = Device.PbDeviceInfo.parseFrom(response.toByteArray())
        return PolarFirmwareVersionInfo(
            deviceFwVersion = devicePbVersionToString(proto.deviceVersion),
            deviceModelName = proto.modelName,
            deviceHardwareCode = proto.hardwareCode
        )
    }

    fun isAvailableFirmwareVersionHigher(currentVersion: String, availableVersion: String): Boolean {
        val current = currentVersion.split(".").map { it.toInt() }
        val available = availableVersion.split(".").map { it.toInt() }

        for (i in current.indices) {
            if (available.size > i) {
                if (current[i] < available[i]) {
                    return true
                } else if (current[i] > available[i]) {
                    return false
                }
            }
        }
        return available.size > current.size
    }

    fun unzipFirmwarePackage(zipBytes: ByteArray): ByteArray {
        try {
            val byteArrayInputStream = ByteArrayInputStream(zipBytes)
            val zipInputStream = ZipInputStream(byteArrayInputStream)
            val byteArrayOutputStream = ByteArrayOutputStream()

            zipInputStream.nextEntry
            val buffer = ByteArray(BUFFER_SIZE)
            var length: Int

            while (zipInputStream.read(buffer).also { length = it } != -1) {
                byteArrayOutputStream.write(buffer, 0, length)
            }

            zipInputStream.closeEntry()
            zipInputStream.close()

            return byteArrayOutputStream.toByteArray()
        } catch (e: Exception) {
            BleLogger.e(TAG, "Failed to unzip firmware package: $e")
            throw e
        }
    }

    private fun devicePbVersionToString(pbVersion: Structures.PbVersion): String
            = "${pbVersion.major}.${pbVersion.minor}.${pbVersion.patch}"
}